File: src/main/java/my/maryvk/maryvkweb/seeker/MarySeekerSchedulerImpl.java
package my.maryvk.maryvkweb.seeker;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ScheduledFuture;

@Service
public class MarySeekerSchedulerImpl implements MarySeekerScheduler {

    @Value("${vk.default-period-to-seek}")
    private int periodToSeek;

    private final MarySeekerFactory marySeekerFactory;

    private final ThreadPoolTaskScheduler taskScheduler;
    private final Map<Integer, ScheduledFuture<?>> scheduledSeekers;

    @Autowired
    public MarySeekerSchedulerImpl(MarySeekerFactory marySeekerFactory) {
        this.marySeekerFactory = marySeekerFactory;

        taskScheduler = new ThreadPoolTaskScheduler();
        taskScheduler.initialize();
        scheduledSeekers = new HashMap<>();
    }

    @Override
    public void schedule(int targetId) {
        if (isRunning(targetId))
            return;
        MarySeeker seeker = marySeekerFactory.create(targetId);
        ScheduledFuture<?> task = taskScheduler.scheduleAtFixedRate(seeker::seek, periodToSeek);
        scheduledSeekers.put(targetId, task);
    }

    @Override
    public void unschedule(int targetId) {
        if (!isRunning(targetId))
            return;
        scheduledSeekers.remove(targetId).cancel(true);
    }

    @Override
    public boolean isRunning(int targetId) {
        return scheduledSeekers.containsKey(targetId);
    }
}

File: src/main/java/my/maryvk/maryvkweb/seeker/MarySeekerSchedulerImpl.kt
package my.maryvk.maryvkweb.seeker

import org.springframework.beans.factory.annotation.Value
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler
import org.springframework.stereotype.Service
import java.util.*
import java.util.concurrent.ScheduledFuture

@Service class MarySeekerSchedulerImpl(
        private val marySeekerFactory: MarySeekerFactory
) : MarySeekerScheduler {

    @Value("\${vk.default-period-to-seek}")
    private val periodToSeek: Long = 0

    private val taskScheduler: ThreadPoolTaskScheduler = ThreadPoolTaskScheduler()
    private val scheduledSeekers: MutableMap<Int, ScheduledFuture<*>>

    init {
        taskScheduler.initialize()
        scheduledSeekers = HashMap<Int, ScheduledFuture<*>>()
    }

    override fun schedule(targetId: Int) {
        if (isRunning(targetId))
            return
        val seeker = marySeekerFactory.create(targetId)
        val task = taskScheduler.scheduleAtFixedRate({ seeker.seek() }, periodToSeek)
        scheduledSeekers.put(targetId, task)
    }

    override fun unschedule(targetId: Int) {
        if (!isRunning(targetId))
            return
        scheduledSeekers.remove(targetId)?.cancel(true)
    }

    override fun isRunning(targetId: Int): Boolean {
        return scheduledSeekers.containsKey(targetId)
    }
}